build.assets/tooling/cmd/changelog/prparser.go

```diff
@@ -0,0 +1,188 @@
+	// Summary is the changelog summary extracted from a PR
+	Summary string
+	Number  int
+	URL     string
+}
+
+const (
+	ossCLTemplate = `
+{{- range . -}}
+* {{.Summary}} [#{{.Number}}]({{.URL}})
+{{ end -}}
+`
+	entCLTemplate = `
+{{- range . -}}
+* {{.Summary}}
+{{ end -}}
+`
+)
+
+var (
+	// clPattern will match a changelog format with the summary as a subgroup.
+	// e.g. will match a line "changelog: this is a changelog" with subgroup "this is a changelog".
+	clPattern = regexp.MustCompile(`[Cc]hangelog: +(.*)`)
+
+	ossCLParsedTmpl = template.Must(template.New("oss cl").Parse(ossCLTemplate))
+	entCLParsedTmpl = template.Must(template.New("enterprise cl").Parse(entCLTemplate))
+)
+
+// pr is the expected output format of our search query
+type pr struct {
+	Body   string `json:"body,omitempty"`
+	Number int    `json:"number,omitempty"`
+	Title  string `json:"title,omitempty"`
+	URL    string `json:"url,omitempty"`
+}
+
+type changelogGenerator struct {
+	isEnt bool
+	dir   string
+}
+
+// dateRangeFormat takes in a date range and will format it for GitHub search syntax.
+// to can be empty and the format will be to search everything after from
+func dateRangeFormat(from, to string) string {
+	if to == "" {
+		return fmt.Sprintf(">%s", from)
+	}
+	return fmt.Sprintf("%s..%s", from, to)
+}
+
+// findChangelog will parse a body of a PR to find a changelog.
+func findChangelog(commentBody string) (found bool, summary string) {
+	// If a match is found then we should get a non empty slice
+	// 0 index will be the whole match including "changelog: *"
+	// 1 index will be the subgroup match which does not include "changelog: "
+	m := clPattern.FindStringSubmatch(commentBody)
+	if len(m) > 1 {
+		return true, m[1]
+	}
+	return false, ""
+}
+
+func prettierSummary(cl string) string {
+	cl = strings.TrimSpace(cl)
+	if !strings.HasSuffix(cl, ".") {
+		cl += "."
+	}
+	return cl
+}
+
+// ghListPullRequests is a wrapper around the `gh` command to list PRs
+// searchQuery should follow the GitHub search syntax
+func (c *changelogGenerator) ghListPullRequests(searchQuery string) (string, error) {
+	var stdout bytes.Buffer
+	var stderr bytes.Buffer
+
+	cmd := exec.Command("gh", "pr", "list", "--search", searchQuery, "--limit", "200", "--json", "number,url,title,body")
+	cmd.Dir = c.dir
+	cmd.Stderr = &stderr
+	cmd.Stdout = &stdout
+
+	err := cmd.Run()
+
+	if err != nil {
+		return strings.TrimSpace(stderr.String()), trace.Wrap(err, "failed to get a list of prs")
+	}
+
+	return strings.TrimSpace(stdout.String()), nil
+}
+
+// parsePRList parses raw output from gh cli
+func parsePRList(data string) ([]parsedPR, error) {
+	parsedList := []parsedPR{}
+
+	// data should be in the format of a list of PR's
+	var list []pr
+	dec := json.NewDecoder(strings.NewReader(data))
+	err := dec.Decode(&list)
+	if err != nil {
+		return parsedList, trace.Wrap(err)
+	}
+
+	for _, p := range list {
+		found, clSummary := findChangelog(p.Body)
+		if found {
```

Comment:
We should only be skipping PRs that have `no-changelog` label set on them. Otherwise, if there's no "changelog:" line in the description, we should take PR title.

Reply:
> we should take PR title

Can we log something when this happs this as well? _Usually_ if a PR is missing the label and an entry then it's worth looking into why this occurred.

Reply:
This is only to catch Enterprise PRs that don't have a changelog. The issue hinted at there not being strict control in the Enterprise repo for the PR to have a `changelog` entry and to not include the title if it isn't there. I can add back the functionality to include a title though.

I could add a section at the bottom for any that do get caught though.

Reply:
> The issue hinted at there not being strict control in the Enterprise repo for the PR to have a changelog entry 

Yea we don't currently, though it'd probably be good to add it if we're going to start including enterprise changelog entries. We have a workflow [here](https://github.com/gravitational/teleport/blob/master/.github/workflows/changelog.yaml) that just needs to get copied to the teleport.e repo.

I can add this to my TODO list.